packages/as-sha256/src/index.ts

```diff
@@ -82,6 +76,163 @@ export function digest64HashObjects(obj1: HashObject, obj2: HashObject): HashObj
   return byteArrayToHashObject(outputUint8Array);
 }
 
+/**
+ * Hash 4 Uint8Array objects in parallel, each 64 length as below
+ * Inputs: 0    1    2    3    4    5    6    7
+ *          \  /      \  /      \  /      \  /
+ * Outputs:  0          1         2         3
+ */
+export function hash4Input64s(inputs: Uint8Array[]): Uint8Array[] {
+  if (inputs.length !== 4) {
+    throw new Error("Input length must be 4");
+  }
+  for (let i = 0; i < 4; i++) {
+    const input = inputs[i];
+    if (input == null) {
+      throw new Error(`Input ${i} is null or undefined`);
+    }
+    if (input.length !== 64) {
+      throw new Error(`Invalid length ${input.length} at input ${i}`);
+    }
+  }
+
+  // set up input buffer for v128
+  inputUint8Array.set(inputs[0], 0);
+  inputUint8Array.set(inputs[1], 64);
+  inputUint8Array.set(inputs[2], 128);
+  inputUint8Array.set(inputs[3], 192);
+
+  ctx.hash4Input64s(wasmOutputValue);
+
+  const output0 = outputUint8Array.slice(0, 32);
+  const output1 = outputUint8Array.slice(32, 64);
+  const output2 = outputUint8Array.slice(64, 96);
+  const output3 = outputUint8Array.slice(96, 128);
+
+  return [output0, output1, output2, output3];
+}
+
+/**
+ * Hash 8 HashObjects in parallel:
+ *   - input${i} has h0 to h7, each 4 bytes which make it 32 bytes
+ *
+ * Inputs      h0    h1    h2    h3    h4    h5    h6   h7
+ *               \   /      \    /       \   /      \   /
+ * Outputs         o0          o1          o2         o3
+ */
+export function hash8HashObjects(inputs: HashObject[]): HashObject[] {
```

Comment:
i think `8` in name is confusing here because hash objects are 4 only each hash object is h0,...h7

Reply:
agree we need to name the 2 new apis better here
- I feel like `digest*()` would be for 1 operation, I propose to make simd apis as `hash*()`
- we have `digest64(data: Uint8Array): Uint8Array`, simd equivalent would be `hash4UintArray64s(inputs: Uint8Array[]): Uint8Array[]` ==> here each input is exactly 64 bytes
- we have `digest64HashObjects(obj1: HashObject, obj2: HashObject): HashObject`, simd equivalent would be `hash4HashObjectInputs(inputs: HashObject[]): HashObject[]` 
  - here `inputs` would be an array of 8 HashObjects (each 2 `HashObjects` is an HashObjectInput). This is easier for the consumer `persistent-merkle-tree` but could be a bit confusing because it's called `hash4*` but input is an array of 8, can add more comments for this
  - other option would be introducing data structure like `{obj1: HashObject; obj2: HashObject}[4]` or `HashObject[2][4]`. This is better in term of clarity but I don't feel like making object creations living really short-tem just for the clarity while the implementation does not need it

let me know what you think @g11tech @wemeetagain

Reply:
i am good with the above suggestion :+1:

Reply:
Maybe call simd functions with "batch" or "simd" in the name. "Digest" vs "hash" is not great because they are mostly used as synonyms.

Reply:
I changed the name to `batchHash*` 👍